src/llanalyzer/Manager.cc

```diff
@@ -0,0 +1,285 @@
+	{
+	return GetComponentTag(name);
+	}
+
+bool Manager::IsEnabled(Tag tag)
+	{
+	if ( ! tag )
+		return false;
+
+	Component* p = Lookup(tag);
+
+	if ( ! p )
+		return false;
+
+	return p->Enabled();
+	}
+
+bool Manager::IsEnabled(EnumVal* val)
+	{
+	Component* p = Lookup(val);
+
+	if ( ! p )
+		return false;
+
+	return p->Enabled();
+	}
+
+Analyzer* Manager::InstantiateAnalyzer(const Tag& tag)
+	{
+	Component* c = Lookup(tag);
+
+	if ( ! c )
+		{
+		reporter->InternalWarning("request to instantiate unknown llanalyzer");
+		return nullptr;
+		}
+
+	if ( ! c->Enabled() )
+		return nullptr;
+
+	if ( ! c->Factory() )
+		{
+		reporter->InternalWarning("analyzer %s cannot be instantiated dynamically", GetComponentName(tag).c_str());
+		return nullptr;
+		}
+
+	Analyzer* a = c->Factory()();
+
+	if ( ! a )
+		{
+		reporter->InternalWarning("analyzer instantiation failed");
+		return nullptr;
+		}
+
+	if ( tag != a->GetAnalyzerTag() )
+		{
+		reporter->InternalError("Mismatch of requested analyzer %s and instantiated analyzer %s. This usually means that the plugin author made a mistake.",
+		                        GetComponentName(tag).c_str(), GetComponentName(a->GetAnalyzerTag()).c_str());
+		return nullptr;
+		}
+
+	return a;
+	}
+
+Analyzer* Manager::InstantiateAnalyzer(const std::string& name)
+	{
+	Tag tag = GetComponentTag(name);
+	return tag ? InstantiateAnalyzer(tag) : nullptr;
+	}
+
+void Manager::ProcessPacket(Packet* packet)
+	{
+#ifdef DEBUG
+	static size_t counter = 0;
+	DBG_LOG(DBG_LLANALYZER, "Analyzing packet %ld, ts=%.3f...", ++counter, packet->time);
+#endif
+
+	if ( ! analyzer_set )
+		return;
+
+	// Dispatch and analyze layers
+	AnalyzerResult result = AnalyzerResult::Continue;
+	identifier_t next_layer_id = packet->link_type;
+	do
+		{
+		auto current_analyzer = analyzer_set->Dispatch(next_layer_id);
+
+		// Analyzer not found
+		if ( current_analyzer == nullptr )
+			{
+			DBG_LOG(DBG_LLANALYZER, "Could not find analyzer for identifier %#x", next_layer_id);
+			packet->Weird("no_suitable_analyzer_found");
+			break;
+			}
+
+		// Analyze this layer and get identifier of next layer protocol
+		std::tie(result, next_layer_id) = current_analyzer->Analyze(packet);
+
+#ifdef DEBUG
+		switch ( result )
+			{
+			case AnalyzerResult::Continue:
+				DBG_LOG(DBG_LLANALYZER, "Analysis in %s succeeded, next layer identifier is %#x.",
+				        current_analyzer->GetAnalyzerName(), next_layer_id);
+				break;
+			case AnalyzerResult::Terminate:
+				DBG_LOG(DBG_LLANALYZER, "Done, last found layer identifier was %#x.", next_layer_id);
+				break;
+			default:
+				DBG_LOG(DBG_LLANALYZER, "Analysis failed in %s", current_analyzer->GetAnalyzerName());
+			}
+#endif
+
+		} while ( result == AnalyzerResult::Continue );
+
+	if ( result == AnalyzerResult::Terminate )
+		CustomEncapsulationSkip(packet);
+
+	// Processing finished, reset analyzer set state for next packet
+	analyzer_set->Reset();
```

Comment:
It's not quite clear to me why the analyzer set is holding the state. It looks like the state is only relevant inside the loop in `ProcessPacket`. Wouldn't it be more natural to hold the state inside that loop then?

Reply:
The `Dispatch()` method depends on the current state. If we keep the separation of Manager and `AnalyzerSet`, this would make sense.